### PR TITLE
Replace deprecated sycl::device_ptr/sycl::host_ptr

### PR DESCRIPTION
**Pull request by masterleinad**
See https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_intel_usm_address_spaces.asciidoc#feature-test-macro. `sycl::ext::intel::device_ptr/host_ptr` was introduced in https://github.com/intel/llvm/pull/7680 and `sycl::device_ptr/host_ptr` was removed in https://github.com/intel/llvm/pull/13240.
This pull request introduces aliases in `Kokkos::Impl::SYCLTypes` based on the level of feature support.

### CODE REVIEW

**Comment by masterleinad** [comment]
The SYCL build passes.

File: core/src/setup/Kokkos_Setup_SYCL.hpp [hunk 1]
```diff
@@ -45,4 +45,21 @@
 #define KOKKOS_IMPL_SYCL_GET_MULTI_PTR(accessor) accessor.get_pointer()
 #endif
 
+// FIXME_SYCL Use type directly once it has stabilized in SYCL.
+namespace Kokkos::Impl::SYCLTypes {
+#ifndef SYCL_EXT_INTEL_USM_ADDRESS_SPACES
+#error SYCL_EXT_INTEL_USM_ADDRESS_SPACES undefined!
+#elif SYCL_EXT_INTEL_USM_ADDRESS_SPACES >= 2
+template <typename T>
+using device_ptr = sycl::ext::intel::device_ptr<T>;
```
Comment (dalg24) [inline core/src/setup/Kokkos_Setup_SYCL.hpp:54]: I would prefer if didn't introduced that nested typename and renamed 
`Kokkos::Impl::SYCLTypes::{host,device}_ptr` to `Kokkos::Impl::sycl_{host,device}_ptr` which I find more readable and consistent with common practice throughout the codebase.

Comment (dalg24) [reply core/src/setup/Kokkos_Setup_SYCL.hpp:54]: It also makes most if not all of the changes in this PR be simply
`sycl::device_ptr -> sycl_device_ptr` and so on.

Comment (masterleinad) [reply core/src/setup/Kokkos_Setup_SYCL.hpp:54]: I don't have a strong opinion. If that gets me your approval, I'm happy to change it.

Comment (masterleinad) [reply core/src/setup/Kokkos_Setup_SYCL.hpp]: See 6333d7d

File: core/src/SYCL/Kokkos_SYCL_Instance.cpp [hunk 2]
```diff
@@ -166,7 +166,7 @@ int SYCLInternal::acquire_team_scratch_space() {
   return current_team_scratch;
 }
 
-sycl::device_ptr<void> SYCLInternal::resize_team_scratch_space(
+Kokkos::Impl::sycl_device_ptr<void> SYCLInternal::resize_team_scratch_space(
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Instance.cpp:169]: You don't need the `Kokkos::Impl::` qualification?
Same question everywhere

**Review by dalg24 — APPROVED** [review]
I would not qualify when not necessary.  I find it unnecessarily verbose.

File: core/src/SYCL/Kokkos_SYCL_Instance.cpp (hunk 2 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Instance.cpp:169]: Yes, I do. I only qualified where necessary. `Kokkos_SYCL_Instance.[cpp|hpp]` implementations are in a `Kokkos::Experimental::Impl` namespace. As soon as we remove `Experimental` from the `SYCL` backend, it wouldn't be necessary anymore.

Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Instance.cpp:169]: Then I need to qualify the other half of the occurrences.

**Comment by masterleinad** [comment]
The SYCL build passes.

**Review by crtrott — APPROVED** [review]

**Comment by ndellingwood** [comment]
Does this merit a changelog entry for 4.4 #6914 ?

**Pull request merged**